core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -172,6 +184,44 @@ public Result applyFailedShards(ClusterState clusterState, List<FailedRerouteAll
         return buildResultAndLogHealthChange(allocation, "shards failed [" + failedShardsAsString + "] ...");
     }
 
+    private ClusterState removeStaleAllocationIds(ClusterState clusterState, List<FailedRerouteAllocation.StaleShard> staleShards) {
+        MetaData oldMetaData = clusterState.metaData();
+        MetaData.Builder metaDataBuilder = null;
+        for (Map.Entry<Index, List<FailedRerouteAllocation.StaleShard>> entry : staleShards.stream().collect(
+            Collectors.groupingBy(fs -> fs.shardId.getIndex())).entrySet()) {
+            final IndexMetaData oldIndexMetaData = oldMetaData.getIndexSafe(entry.getKey());
+            IndexMetaData.Builder indexMetaDataBuilder = null;
+            for (FailedRerouteAllocation.StaleShard staleShard : entry.getValue()) {
+                // remove allocation ids for stale shards
+                int shardNumber = staleShard.shardId.getId();
+                Set<String> inSyncAllocations = indexMetaDataBuilder == null ?
+                    oldIndexMetaData.inSyncAllocationIds(shardNumber) :
+                    indexMetaDataBuilder.getInSyncAllocationIds(shardNumber);
+                String shardAllocationId = staleShard.allocationId;
+                if (inSyncAllocations.contains(shardAllocationId) && inSyncAllocations.size() > 1) {
```

Comment:
I think some comments here would help. Why only if there is more than one in sync allocation? Is that because we only care about allocation ids if there is more than just the primary?